connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
+                    jsonDocuments[0] = entity.getBody();
+                });
+        Assert.isTrue(Objects.nonNull(jsonDocuments[0]), "JsonDocument should not be null.");
+
+        final List<String> comments = jsonDocuments[0].read("$.values[*].comment.text");
+        return comments.stream()
+                .limit(COMMENTS_SIZE)
+                .collect(Collectors.toList());
+    }
+
+    private Single<ResponseEntity<String>> performStashAction(final String baseUrl,
+                                                              final String authHeader,
+                                                              final StashPullRequest pullRequest,
+                                                              final StashAction stashAction) {
+        final HttpHeaders headers = buildHeaders(authHeader);
+
+        // Get current version of the pull request. Pull request "version" changes when we do any actions on it.
+        // When the pull request is raised, the current value will be 0.
+        // For example, when we approve the pull request, then the version will change from 0 to 1.
+        // We have to add the latest version of the pull request URI to do any actions. Otherwise, the action will be rejected.
+        // If the build for the branch is going on, then the actions would be rejected.
+        final String version = getVersion(headers, baseUrl, pullRequest);
+
+        final String url = String.format(StashConstants.STASH_PR_ACTION_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId(),
+                stashAction.getAction(),
+                version);
+
+        // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request action.
+        headers.add(ATLASSIAN_TOKEN, "no-check");
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private HttpHeaders buildHeaders(final String authHeader) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+        return headers;
+    }
+
+    private String getVersion(final HttpHeaders headers,
+                              final String baseUrl,
+                              final StashPullRequest pullRequest) {
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = getPullRequestInfo(headers, pullRequest, baseUrl);
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
```

Comment:
Again, I suspect this won't work.  My reading of the code would expect this to asynchronously fall through and hit the assert before subscribe is called (at least on a real system -- unit test frameworks/libraries/etc. tend to make asynchronous code behave synchronously).

Reply:
Change this to something like:

```java
    private Single<String> getVersion(final HttpHeaders headers,
                              final String baseUrl,
                              final StashPullRequest pullRequest) {
        return getPullRequestInfo(headers, pullRequest, baseUrl)
                .map(ResponseEntity::getBody)
                .doOnSuccess(jsonDocument -> logger.debug("getVersion returned, jsonDoc={}", jsonDocument))
                .map(jsonDocument -> Integer.toString(jsonDocument.read("$.version")));
    }
```

```java
        return getVersion(headers, baseUrl, pullRequest)
                .flatMap(version -> {
                    final String url = String.format(StashConstants.STASH_PR_ACTION_FORMAT,
                            baseUrl,
                            pullRequest.getProjectKey(),
                            pullRequest.getRepositorySlug(),
                            pullRequest.getPullRequestId(),
                            stashAction.getAction(),
                            version);

                    // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request action.
                    headers.add(ATLASSIAN_TOKEN, "no-check");
...
```

Reply:
I have changed the code to above format..